ruby/ql/lib/codeql/ruby/experimental/UnicodeBypassValidationQuery.qll

```diff
@@ -0,0 +1,91 @@
+/**
+ * Provides a taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ */
+
+private import ruby
+private import codeql.ruby.dataflow.RemoteFlowSources
+private import codeql.ruby.Concepts
+private import codeql.ruby.TaintTracking
+private import codeql.ruby.ApiGraphs
+import UnicodeBypassValidationCustomizations::UnicodeBypassValidation
+
+/** A state signifying that a logical validation has not been performed. */
+class PreValidation extends DataFlow::FlowState {
+  PreValidation() { this = "PreValidation" }
+}
+
+/** A state signifying that a logical validation has been performed. */
+class PostValidation extends DataFlow::FlowState {
+  PostValidation() { this = "PostValidation" }
+}
+
+/**
+ * A taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ *
+ * This configuration uses two flow states, `PreValidation` and `PostValidation`,
+ * to track the requirement that a logical validation has been performed before the Unicode Transformation.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnicodeBypassValidation" }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    source instanceof RemoteFlowSource and state instanceof PreValidation
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node nodeFrom, DataFlow::FlowState stateFrom, DataFlow::Node nodeTo,
+    DataFlow::FlowState stateTo
+  ) {
+    (
+      exists(Escaping escaping | nodeFrom = escaping.getAnInput() and nodeTo = escaping.getOutput())
+      or
+      exists(RegexExecution re | nodeFrom = re.getString() and nodeTo = re)
+      or
+      // String Manipulation Method Calls
+      // https://ruby-doc.org/core-2.7.0/String.html
+      exists(DataFlow::CallNode cn |
+        cn.getMethodName() =
+          [
+            [
+                "ljust", "lstrip", "succ", "next", "rjust", "capitalize", "chomp", "gsub", "chop",
```

Comment:
I don't understand why many of these string manipulations should switch the state from `PreValidation` to `PostValidation`. Shouldn't they be state preserving?

Reply:
That is what I thought initially, but:

1. replace - sub - start_with - match are okey. 
2. In a similar way: end_with - gsub
3. lstrip - strip
```ruby
s="                     sim4n6                     "
s2=s.lstrip().unicode_normalize(:nfkc)
print(s2)
print(s2.include?(" "))
```
4. tr_s - tr
```ruby
print("hayｋing".tr_s('k', 'r').unicode_normalize(:nfkc).include?("k"))
```
5. index - index
```ruby
i1 = "ᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱ".rindex("E")
i2 = "ᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱᴱ".unicode_normalize(:nfkc).rindex("E")
print(i1, "---", i2)
```
6. delete - delete_suffix - delete_prefix
```ruby
"﹤".delete("<").unicode_normalize(:nfkc).include?("<") => true
```
7. include
```ruby
s="﹤"
if !s.include?("<")
	s2 = s.unicode_normalize(:nfkc)
	print(s2.include?("<"))
end
```

Reply:
I would provide further examples in the next comment to validate as much as possible with examples ...

Reply:
8. partition - rparition

```ruby
s1 = "heˡˡo".partition("l")
s2 = "heˡˡo".unicode_normalize(:nfkc).partition("l")
print(s1, "----", s2)
```

9. squeeze
```ruby
s1 = "yelˡow mºon".squeeze
s2 = "yelˡow mºon".unicode_normalize(:nfkc).squeeze
print(s1, "----", s2)
```

Reply:
10. split
 
 ```ruby
str = "This can be split into smaller pieces"
print str.split(" ")
print str.unicode_normalize(:nfkc).split(" ")
```

11. count
```ruby
str = "  "
print str.count(" ")
print str.unicode_normalize(:nfkc).count(" ")
```

Reply:
12. scan

```ruby
phrase = "Hello World! This is a sample phrase."

# Define the regular expression pattern
pattern = /(.+?)\s(.+?)\s(.+)/

# Use the scan method with the pattern to split the phrase
matches = phrase.scan(pattern)
matches2 = phrase.unicode_normalize(:nfkc).scan(pattern)

# Extract the three parts from the matches array
if matches.any?
  part1 = matches[0][0]
  part2 = matches[0][1]
  part3 = matches[0][2]

  puts "Part 1: #{part1}"
  puts "Part 2: #{part2}"
  puts "Part 3: #{part3}"
end

print "\n\n"
if matches2.any?
  part1 = matches2[0][0]
  part2 = matches2[0][1]
  part3 = matches2[0][2]

  puts "Part 1: #{part1}"
  puts "Part 2: #{part2}"
  puts "Part 3: #{part3}"
end
```

Reply:
13. downcase - swapcase - uprcase

```ruby
puts "℻".downcase.unicode_normalize(:nfkc)
puts "℻".unicode_normalize(:nfkc).downcase
```

14. succ - next

```ruby
s = "Ⅳ"
puts s.succ.unicode_normalize(:nfkc)
puts s.unicode_normalize(:nfkc).succ
```

Reply:
Thanks for writing up such detailed examples! I see, I got to learn a lot about unicode ;-)

Reply:
There are always surprises with Unicodes :) 

I see that you merged the PR. Do you still want me to look for potential Unicode characters for the rest of the String Manipulation methods, please ...?

Reply:
No need to provide examples for all of them. Perhaps we should add a small paragraph to the qhelp documentation and include a convincing example with one of the string manipulation methods. Ideally with a commonly used method that unexpectedly causes an obvious problem.

Reply:
I love this one ... 

> "﹤".delete("<").unicode_normalize(:nfkc).include?("<")

I am adding it right away.

Reply:
and like always they may be some false positives.